### ISSUE TITLE
VA Memorable Date Web Component validation faltering, reporting "XX" as day

### ISSUE THREAD

**Issue by zack** [department-of-veterans-affairs/vets-design-system-documentation#2283]
# Bug Report

- [x] I’ve searched for any related issues and avoided creating a duplicate issue.

## What happened

Code here: https://github.com/department-of-veterans-affairs/vets-website/blob/0e6deba8c8c32aabc9b46048a62cd9134296e489/src/applications/disability-benefits/all-claims/pages/reservesNationalGuardService.jsx#L28

Using the [VA Memorable Date Component](https://design.va.gov/storybook/?path=/docs/uswds-va-memorable-date--default), a user is able to force an invalid date to be accepted by entering non-numeric characters in the day field then clicking around the page. The validation error that initially arises eventually clears itself and the component reports "XX" for the day, which is causing form submissions to fail. There is a PR open reverting use of the component which shows some screenshots of the issue [here](https://design.va.gov/storybook/?path=/docs/uswds-va-memorable-date--default).

## What I expected to happen

I expect the component to force the user to enter a valid date.

## Reproducing
- Formation version:
- Device: MacOS v13.6
- Browser: Firexox Developer Edition 121.0b3

Steps to reproduce:

1. Start form VA Form 21-526EZ
2. Add Coast Guard Reserves to military service history
3. On the page for Reserve and National Guard Information:
  1. Select a valid month from the dropdown
  2. Enter a valid year
  3. Enter "AA" for the day - notice an error
  4. Click into unit name - notice the error persists
  5. Click on the page background - notice the error disappears

https://github.com/department-of-veterans-affairs/vets-design-system-documentation/assets/2482403/a4b1f49f-c253-41bf-a533-b8e112abff0f

## Urgency
How urgent is this request? Please select the approriate option below and/or provide details
- [ ] This bug is blocking work currently in progress
- [x] This bug is affecting work currently in progress but we have a workaround
- [ ] This bug is blocking work planned within the next few sprints
- [ ] This bug is not blocking any work
- [ ] Other

### Details

For the time being we are reverting to use an older component for the date.

**Comment by jamigibbs** [comment]
Slack support convo: https://dsva.slack.com/archives/CBU0KDSB1/p1701277278795159

**Comment by caw310** [comment]
Hey team! Please [add your planning poker estimate](https://app.zenhub.com/workspaces/Platform-Design-System-5f8de67192551b0012ebb802/issues/department-of-veterans-affairs/vets-design-system-documentation/2283?planning-poker) with Zenhub @Andrew565 @ataker @harshil1793 @it-harrison @jamigibbs @micahchiang @nickjg231 @powellkerry @rmessina1010 @rsmithadhoc

**Comment by caw310** [comment]
Validate the source of the problem. IF it is DS issue, DST will fix.

**Comment by nickjg231** [comment]
I do not believe this is possible anymore as the Memorable Date defaults to using a dropdown for Day, so the user is unable to enter AA for the day <img width="948" src="https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBMGFxQWc9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--25e7c59cbaea99c941e04b580ce4636363a4bef8/image.png" alt="image.png" />

**Comment by jamigibbs** [comment]
@nickjg231 Do you mind double checking that? When I'm on Chrome, I can manually enter a value into the inputs on both the USWDS v3 and v1 versions:

![Screenshot 2024-01-09 at 8 49 29 AM](https://github.com/department-of-veterans-affairs/vets-design-system-documentation/assets/872479/aea49a26-4c7e-476e-9bbb-b2c7443a4e2d)

![Screenshot 2024-01-09 at 8 48 32 AM](https://github.com/department-of-veterans-affairs/vets-design-system-documentation/assets/872479/f91aa25c-31c4-4665-8c41-0811eaa9f33a)

**Comment by micahchiang** [comment]
@jamigibbs @nickjg231 - I wasn't on standup so apologies if I missed an update on this. Was it verified that dropdown usage prevents manually entering values into the day field?

**Comment by micahchiang** [comment]
@nickjg231 - I'm reopening this, as the work still needs to be investigated. The issue persists on Chrome and Firefox. Please look into this further to determine if it's something that needs to be addressed in our memorable date component, or something that needs to be fixed with how the Forms Library is managing form state. Thanks!

**Comment by nickjg231** [comment]
I was incorrect when I said this was `VaMemorableDate`. Upon further investigation, the `Reserve and National Guard Information` page is defined in the following file:
`src/applications/disability-benefits/all-claims/pages/reservesNationalGuardService.jsx`

In this file, the date fields are tied to the date range `obligationTermOfServiceDateRange`, where both the `from` and `to` properties in the range are rendered with a `dateUI` Forms System component (from `@department-of-veterans-affairs/platform-forms-system/date`):
<img width="498" src="https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBMG5RQWc9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--8ca1eb0aefc5e05b006e0914e5c5d05d7a6a1d21/image.png" alt="image.png" />

When looking on the page itself in staging, the fields are a fields with select inputs, not a VaMemorableDate:
 <img width="2000" src="https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBMHZRQWc9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--d0cda31a3d2dbf9837da551e0f3d39bd25a2fd1a/image.png" alt="image.png" />

It does not use text inputs for the day field as stated in the issue initially, so I do not believe this error should exist anymore. However, this is not something that needs to be addressed in our components regardless.

**Comment by nickjg231** [comment]
This will be carrying over for further investigation

**Comment by nickjg231** [comment]
Upon reviewing the issue with the `VAMemorableDateField` web-component, I've identified a couple of key areas that require adjustments in the current implementation, which I believe will resolve the validation issues the user was having:

- Validator: The implementation in the GitHub blob uses the `validateDate` validator. However, this validator is not built to be used with the `VaMemorableDateField` web-component. The correct validator to use in this context is `validateCurrentOrPastMemorableDate`. This is something I confirmed with Robert on the forms team. Implementing this specific validator should effectively resolve the validation issues the user had encountered. When testing a form with the `validateDate` validator and the `VaMemorableDateField` web-component locally, I was able to reproduce a similar sort of validation error state as the original issue (though I could not reproduce exactly the same error). In my case, the day field was sometimes marked as valid even when it had something like “AB” entered, but it would be correctly changed to an error upon further interaction with an input on the page.

- Property Definition Refinement: The current approach of defining the `from` and `to` properties manually can be streamlined for efficiency. Instead of the existing method, it's recommended by the forms team to use a pattern like `from: currentOrPastDateUI`. This pattern encompasses the necessary validator, error messages, and other relevant properties, tailored for compatibility with the web component. It also permits customizing properties like `title` still. It's important to note that `CurrentOrPastDateUI` should be imported from `platform/forms-system/src/js/web-component-patterns` as this is the newer version of the class.

In conclusion, I believe these changes will fix the validation issue encountered by the user. Given that these components and validators exist in the Forms Library, the Design System Team should not need to make any updates.
